File: backend/app/models/follow_up.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Date, Time
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database import Base


class FollowUp(Base):
    __tablename__ = "follow_ups"
    
    id = Column(Integer, primary_key=True, index=True)
    
    # Related lead
    lead_id = Column(Integer, ForeignKey("leads.id"), nullable=False)
    
    # Scheduling
    scheduled_date = Column(Date, nullable=False)
    scheduled_time = Column(Time, nullable=True)
    
    status = Column(String(50), default="Pending")  # Pending, Completed, Cancelled
    outcome = Column(Text, nullable=True)  # Result of the follow-up
    notes = Column(Text, nullable=True)
    
    # Completed
    completed_at = Column(DateTime, nullable=True)
    
    # Owner
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    
    # Relationships
    lead = relationship("Lead", back_populates="follow_ups")
    created_by = relationship("User", back_populates="follow_ups")

File: flutter_app/lib/data/models/follow_up.dart
class FollowUp {
  final int id;
  final int leadId;
  final String? leadName;
  final String? leadCompany;
  final String status;
  final String? notes;
  final String? outcome;
  final DateTime? scheduledDate;
  final String? scheduledTime;
  final DateTime? completedAt;
  final DateTime? createdAt;

  const FollowUp({
    required this.id,
    required this.leadId,
    this.leadName,
    this.leadCompany,
    required this.status,
    this.notes,
    this.outcome,
    this.scheduledDate,
    this.scheduledTime,
    this.completedAt,
    this.createdAt,
  });

  bool get isOverdue =>
      status.toLowerCase() != 'completed' &&
      scheduledDate != null &&
      scheduledDate!.isBefore(DateTime.now().subtract(const Duration(hours: 12)));

  bool get isDueToday {
    if (scheduledDate == null || status.toLowerCase() == 'completed') return false;
    final now = DateTime.now();
    return scheduledDate!.year == now.year &&
        scheduledDate!.month == now.month &&
        scheduledDate!.day == now.day;
  }

  factory FollowUp.fromJson(Map<String, dynamic> json) => FollowUp(
        id: json['id'] as int,
        leadId: json['lead_id'] as int,
        leadName: json['lead_name'] as String?,
        leadCompany: json['lead_company'] as String?,
        status: json['status']?.toString() ?? 'Pending',
        notes: json['notes'] as String?,
        outcome: json['outcome'] as String?,
        scheduledDate: json['scheduled_date'] != null
            ? DateTime.tryParse(json['scheduled_date'].toString())
            : null,
        scheduledTime: json['scheduled_time'] as String?,
        completedAt: json['completed_at'] != null
            ? DateTime.tryParse(json['completed_at'].toString())
            : null,
        createdAt: json['created_at'] != null
            ? DateTime.tryParse(json['created_at'].toString())
            : null,
      );
}
